addons/ondevice-controls/tsconfig.json

```diff
@@ -1,13 +1,12 @@
 {
   "compilerOptions": {
     "baseUrl": ".",
-    "jsx": "react-native",
+    "jsx": "react",
     "rootDir": "./src",
-    "module": "ESNext",
+    "module": "ES6",
     "skipDefaultLibCheck": true,
     "skipLibCheck": true,
-    "target": "ESNext",
-    "lib": ["ESNext"],
```

Comment:
This is a bit complex and I'm not 100% sure that I have understood it correctly, but here's how I think`target` and `lib` work:

* `target`: Defines the JavaScript syntax that is generated (e.g. `ES2016`)
* `lib`: What level of JavaScript is the runtime, _including any polyfills you might bundle in_, expected to support.

If `lib` isn't set, it defaults to the same as the `target` (`ES2016` in this case). The possibility to set a different `target` and `lib` seems to be a way to get around situations where you're including polyfills for some new APIs but want to in general compile down to an older syntax.

https://stackoverflow.com/a/42097465/843984 is a pretty good answer on this. The linked TSConfig docs are also worth a read.

Reply:
The way I thought it works is that lib just defines what code you allow in your codebase, such that if you use ESNext then you can use esnext features in the code base but then once you compile it it gets tranformed to the target javascript. So based on that there is no reason to remove the lib option unless we want to change the type of code we write on the project.

Reply:
Using `"lib": ["ESNext"],` seems to allow using ESNext features which get bundled into the compiled code. One way to test this is to do the following:

1. Change `addons/ondevice-controls/tsconfig.json` to include `"lib": ["ESNext"],` in addition to the `"target": "ES2016",`
1. In `addons/ondevice-controls/src/ControlsPanel.tsx`, add `Promise.any([Promise.reject(1), Promise.resolve(2)]).then((it) => console.log(`resolved ${it}`));` to the component
1. Run `yarn prepareAll` in the project root. It should compile without issues.
1. Open `addons/ondevice-controls/dist/ControlsPanel.js`. The code that uses the ESNext feature `Promise.any` should be included in the compiled source.
1. Set up a project that refers to these local sources. If the runtime does not support `Promise.any` (e.g. Node 14 and below), running the project should fail with 

```
TypeError: Promise.any is not a function. (In 'Promise.any([Promise.reject(1), Promise.resolve(2)])', 'Promise.any' is undefined)

This error is located at:
    in ControlsPanel (created by Wrapper)...
```

![Screen Shot 2021-09-12 at 0 09 25](https://user-images.githubusercontent.com/6605505/132961682-b4a0760d-5f88-47fb-8914-8a844d0c8e0c.png)

If you remove `"lib": ["ESNext"],` from `addons/ondevice-controls/tsconfig.json` and run `yarn prepareAll` again, then compile should fail with

```
src/ControlsPanel.tsx(47,11): error TS2550: Property 'any' does not exist on type 'PromiseConstructor'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2021' or later.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Reply:
So what should we do? go ahead with lib ESNext?

Reply:
Would remove the lib and just go with the `target` setting. To keep the codebase modern, we can check how aggressive we could be with the `target` setting without breaking runtime environments. It might be possible to e.g. bump from `ES6`/`ES2015` all the way to `ES2019`, but I haven't tested that.

Reply:
I've checked out this PR and it seems to work well. I'm happy to merge this and we'll see how it goes and we can update it later if necessary.